File: iOS/Puzzlr/Hyperledger/UserBlockchain.swift
//
//  UserBlockchain.swift
//  Puzzlr
//
//  This class represent the Blockchain containing user pseudonym and password.
//
//  Created by Quentin Le Sceller on 09/03/2016.
//  Copyright © 2016 Quentin Le Sceller. All rights reserved.
//

import Foundation

class UserBlockchain {
    
    // Var used to know if the user is connected.
    var connected : Bool
    
    // Use to store variable on the iDevice.
    let defaults = NSUserDefaults.standardUserDefaults()
    
    // The blockchain main object.
    let blockchain : OpenBlockchain
    
    // The language used for the chaincode.
    let type : String
    
    // The name of the chaincode (usually hash value).
    let name : String
    
    // Initializer
    init(ip: String,port : Int) {
        blockchain = OpenBlockchain(ip: ip, port: port)
        type = "GOLANG";
        name = "e72785e0db781f0f37914d5279e26cb001c098fc1f79dfad1220a3e780ba253fea6703755556d072aeb57a939fe9804ee1d80037f41534bb29718da3978e977e";
        
        if defaults.boolForKey("connected") == Bool(true) {
            connected = defaults.boolForKey("connected")
        }
        else {
            connected = Bool(false)
        }
        
    }
    
    
    /**
     Get if an user is registered
     
     - parameter user: the username
     - parameter sync: synchronous or asynchronous request
     
     - returns: true if the user is registered, false otherwise:
     */
    func getRegistered(user : String, sync : Bool) -> Bool {
        var registered = Bool(false)
        let args : [String] = [user]
        
        let response = blockchain.query(type, name: name, function: "query", args: args, sync: sync)
        
        if response.containsString("isRegisteredTrue"){
            registered = Bool(true)
        }
        return registered
    }
    
    /**
     Register an user
     
     - parameter user:     the username
     - parameter password: the password
     - parameter sync:     synchronous or asynchronous request
     
     - returns: true if successfully registered, false otherwise.
     */
    func registerUser(user : String, password : String, sync : Bool) -> Bool {
        var success = Bool(true)
        let args : [String] = [user,password]
        
        if !getRegistered(user, sync: sync) {
            blockchain.invoke(type, name: name, function: "invoke", args: args, sync: sync)
        } else {
            success = Bool(false)
        }
        return success
    }
    
    
    /**
     Delete an user from the database.
     
     - parameter user:     the username
     - parameter password: the password
     - parameter sync:     synchronous or asynchronous request
     
     - returns: true if successfully deleted, false otherwise.
     */
    func deleteUser(user : String,password : String, sync : Bool) -> Bool {
        var success = Bool(true)
        let args : [String] = [user]
        
        if login(user, password: password, sync: sync) {
            blockchain.invoke(type, name: name, function: "delete", args: args, sync: sync)
        } else {
            success = Bool(false)
        }
        
        return success
    }
    
    /**
     Log the user:
     
     - parameter user:     the username
     - parameter password: the password
     - parameter sync:     synchronous or asynchronous request
     
     - returns: true if successfully logged, false otherwise.
     */
    func login(user : String, password : String, sync : Bool) -> Bool {
        var retrievedHashedPassword : NSString?
        var success = Bool(false)
        let args : [String] = [user,""]
        
        // Checking if the user is registered.
        if getRegistered(user, sync: sync){
            
            let result = blockchain.query(type, name: name, function: "query", args: args, sync: sync)
            retrievedHashedPassword = result.substringFromIndex(7)
            let index = retrievedHashedPassword!.length.advancedBy(-2)
            retrievedHashedPassword = retrievedHashedPassword!.substringToIndex(index)
            if (JKBCrypt.verifyPassword(password, matchesHash: String(retrievedHashedPassword!)) == Bool(true)) {
                success = Bool(true)
            }
        }
        return success
    }
    
    
}

File: Android/Puzzlr/app/src/main/java/BlockChain/UserBlockchain.java
package BlockChain;

/**
 * Created by aniss on 09/03/16.
 */

import android.widget.Toast;

import security.BCrypt;

/**
 * The Class UserBlockchain.
 *
 * @author Quentin Le Sceller
 */
public class UserBlockchain implements IUserBlockchain {

    /** The blockchain. */
    OpenBlockchain blockchain;

    /** The type. */
    String type;

    /** The name. */
    String name;

    /**
     * Instantiates a new user blockchain.
     *
     * @param address
     *            the address
     * @param port
     *            the port
     */
    public UserBlockchain(String address, int port) {
        blockchain = new OpenBlockchain(address, port);
        type = "GOLANG";
        name = "622fa47b0be4cffc3a7e913e72622abe6b224caf83bac5df61bba599e33ad291f3d730c31583ba404d9ad03e91d59f15659f3ff326dc30f71a6f0c12b35ed924";
    }

    /*
     * (non-Javadoc)
     *
     * @see userblockchain.IUserBlockchain#getRegistered(java.lang.String)
     */
    @Override
    public boolean getRegistered(String user) {
        boolean registered = false;
        String[] args = { user };

        JSONObject registeredJSON = blockchain.query(type, name, "query", args);

        try {
            if (registeredJSON.getString("OK").contains("isRegisteredTrue")) {
                registered = true;
            }

        } catch (JSONException e) {

            e.printStackTrace();
        }



        return registered;
    }


    /*
     * (non-Javadoc)
     *
     * @see userblockchain.IUserBlockchain#registerUser(java.lang.String,
     * java.lang.String)
     */
    @Override
    public boolean registerUser(String user, String hash) {
        boolean success = true;
        String[] args = { user, hash };

        if (!getRegistered(user)) {
            blockchain.invoke(type, name, "invoke", args);
        } else {
            success = false;
        }

        return success;
    }

    /*
     * (non-Javadoc)
     *
     * @see userblockchain.IUserBlockchain#deleteUser(java.lang.String,
     * java.lang.String)
     */
    @Override
    public boolean deleteUser(String user, String hash) throws JSONException {
        boolean success = true;
        String[] args = { user };

        if (login(user, hash)) {
            blockchain.invoke(type, name, "delete", args);
        } else {
            success = false;
        }

        return success;
    }

    /*
     * (non-Javadoc)
     *
     * @see userblockchain.IUserBlockchain#queryHash(java.lang.String,
     * java.lang.String)
     */
    @Override
    public boolean login(String user, String password) throws JSONException {
        boolean success = false;
        String[] args = { user, password };


        if (getRegistered(user)) {
            JSONObject result;


            result = blockchain.query(type, name, "query", args);

            String retrievedHashedPassword = result.getString("OK");

            BCrypt bCrypt = new BCrypt();



            if(bCrypt.checkpw(password, retrievedHashedPassword)){

                success = true;
            }





        }

        return success;
    }

}